app/routes.py

```diff
@@ -1 +1,167 @@
-from flask import Blueprint
\ No newline at end of file
+from flask import Blueprint, request, jsonify
+from app import db
+from app.models.task import Task
+from app.models.goal import Goal
+from app.helper_functions import validate_model, query_sort, filter_results, filter_and_sort, send_slack_message
+from datetime import datetime
+
+tasks_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+goals_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+@tasks_bp.route("", methods=["GET"])
+def get_all_tasks():    
+    tasks = filter_and_sort(Task)
+    
+    tasks_response = [task.to_json() for task in tasks]
+    
+    return jsonify(tasks_response), 200
+
+# Route to get task by task id
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def get_task_by_id(task_id):
+    task = validate_model(Task, task_id)
+
+    return {"task": task.to_json()}, 200
+
+# Route to create new task
+@tasks_bp.route("", methods=["POST"])
+def create_task():
+    new_task = Task.from_json(request.get_json())
+    
+    db.session.add(new_task)
+    db.session.commit()
+
+    return {"task": new_task.to_json()}, 201
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+    request_body = request.get_json()
+    
+    task.title = request_body["title"],
+    task.description = request_body["description"]
+    
+    db.session.commit()
+
+    return {"task": task.to_json()}, 200
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_model(Task, task_id)
+
+    db.session.delete(task)
+    db.session.commit()
+
+    return {
+        "details": f'Task {task.id} "{task.title}" successfully deleted'
+    }, 200
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = validate_model(Task, task_id)
+
+    task.completed_at = None
+
+    db.session.commit()
+
+    return {"task": task.to_json()}, 200
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = validate_model(Task, task_id)
+
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+    
+    send_slack_message(Task, task)
```

Comment:
Nice work on having your Slack post sent after the logic of marking a task complete! We don't want to send out any false positive alerts just in case our logic fails during the update!